fix: allow IN queries on __key__

danieljbruce opened the pull request:
The fix contains:

- An update to the way that entities are encoded
- An update to tests so that an instanceOf check passes
- A cleanup of the PropertyFilter function
- An additional test to check for IN with keys

Fixes #1024

danieljbruce committed:
Change to try to ignore __key__ as it is redundant

This change matches the change shown in another PR, but the test cases won’t allow for the new change. There must be some edge cases that haven’t been considered here.

danieljbruce committed:
Add a check for isDsKey

isDsKey currently does not check to see that the key is an entity.Key properly so we need to modify this function so that it does a proper check.

danieljbruce committed:
Revert "Add a check for isDsKey"

This reverts commit 4f0db50bb26e54ac2c9b60d53beec6993d244de9.

danieljbruce committed:
Add comments so that key is never reassigned

danieljbruce committed:
Revert "Add comments so that key is never reassigned"

This reverts commit d2795951ac26375ea1c878f041a5952f3572f706.

danieljbruce committed:
Use global entity to test

For an instanceof check to pass, we need to import the entity property the right way

danieljbruce committed:
clean up the property class

Make code a lot more succinct for the PropertyFilter class before adding the new test in.

danieljbruce committed:
Inline encodeValue

There is no need for a private encodeValue function so we just inline this change.

danieljbruce committed:
Add a test for supporting key with IN

This test supports using __key__ with IN and an array.

danieljbruce committed:
Modifies the query proto to correct the test

This change modifies the proto to be what we want it to be so that the deep strict equal check works.

danieljbruce committed:
Remove comments and clean up test

This commit removes a comment that isn’t necessary anymore and renames a variable to demonstrate its proper use.

danieljbruce committed:
Use global entity in the new test instead of entit

The new test doesn’t enter the DSKey check when the entity is used from the beforeEach hook. Use the global entity instead.

danieljbruce committed:
Move the query proto into the test

The specific query proto used should be moved into the test as it is only used by that specific test.

vicb commented on src/filter.ts:
```diff
@@ -13,7 +13,7 @@
 
 import {Operator, Filter as IFilter} from './query';
-import {entity} from './entity';
+import {entity, ValueProto} from './entity';
```
Please fix (remove `ValueProto`)

danieljbruce replied:
Done

vicb commented on src/filter.ts:
```diff
@@ -100,11 +81,7 @@ export class PropertyFilter extends EntityFilter implements IFilter {
-      this.op,
-      this.val
-    ).encodedValue();
+    const value = entity.encodeValue(this.val, this.name);
```
You can also `value: entity.encodeValue(this.val, this.name)` inline. Not sure what the preferred style is.

danieljbruce replied:
Yes. This can be even more compact.

vicb reviewed (approved):
Looks nice 👍 

There are a few "GitHub Actions / lint" that look worth addressing - IMO the typings of this lib could use some love but out of scope for this PR.

danieljbruce commented on test/entity.ts:
```diff
@@ -1894,8 +1894,62 @@ describe('entity', () => {
       assert.deepStrictEqual(entity.queryToQueryProto(query), queryProto);
     });
 
+    it('should support using key with IN', () => {
```
Good point.

I think more specifically, it is addressing the case of using an array as a value as that was what was causing the error in the first place.

danieljbruce committed:
PR updates

Remove an unused import, make value more inline and change name of test case to reflect new functionality.

danieljbruce commented:
@vicb There is just one more thing we should confirm. I notice that `val` in Property Filter is of type `any` so even though filtering by `__key__` ought to use only values that are keys and arrays of keys, the compiler doesn't guarantee it. Are these definitely the only two cases with which we can expect users to filter with `__key__`?

danieljbruce committed:
Added a system test to validate __key__ with array

Since the new test simply introduces a new proto and checks it against __key__ with an array, we still have no way to verify if this encoding will actually work. In this system test we verify that using __key__ with an array actually works in a query.

danieljbruce committed:
use unknown instead of any for value

Modify the type to use `unknown` instead of `any`.

danieljbruce committed:
Introduce restricted value when key is used

Leverage typescript to let the user know at compile time that Keys and arrays of keys are to be used when __key__ is used.

danieljbruce committed:
Add the restricted value type to query too

The restricted value type should be used on filter as well as the property filter.

vicb commented on src/filter.ts:
```diff
@@ -56,42 +57,34 @@ export abstract class EntityFilter {
   abstract toProto(): any;
 }
 
+export type restrictedValueType<T> = T extends '__key__'
```
`RestrictedValueType`

vicb commented on src/query.ts:
```diff
@@ -207,12 +207,19 @@ class Query {
+    value?: restrictedValueType<T> | null
+  ): Query;
+  filter<T extends string>(
+    propertyOrFilter: T,
```
T | EntityFilter

vicb commented on src/query.ts:
```diff
@@ -207,12 +207,19 @@ class Query {
+  ): Query;
+  filter<T extends string>(
+    propertyOrFilter: T | EntityFilter,
     operatorOrValue?: Operator,
```
This should be `restrictedValueType<T> | null | Operator`

danieljbruce replied:
Signature has been updated

danieljbruce commented:
I mean see https://github.com/googleapis/nodejs-datastore/pull/1089/commits/8d269d2e2cd60bef27ff6856bffe21c5a8f9f20c

danieljbruce commented:
I mean https://github.com/googleapis/nodejs-datastore/pull/1089

danieljbruce committed:
Modify the interface for filter

Checking for nulls is now included in the restricted value type so we don’t need to declare it in the signature of the function.

danieljbruce committed:
rename to allowedFilterValueType

rename a value type variable to allowed filter value type to more accurately describe the type’s purpose.

danieljbruce committed:
Use capital letter to follow name conventions

The old variable name doesn’t follow the naming convention.

danieljbruce committed:
type guard should check instance of

prettier and change the typeguard so that it checks to see if the input is an instance of EntityFilter.

danieljbruce committed:
Merge branch 'main' into 1065-query-filter-for-IN-with-keys-test

kolea2 commented on src/filter.ts:
```diff
@@ -156,5 +144,5 @@ class CompositeFilter extends EntityFilter {
 
 export function isFilter(filter: any): filter is EntityFilter {
-  return (filter as EntityFilter).toProto !== undefined;
+  return filter instanceof EntityFilter;
```
why does this have to change?

vicb replied:
I suggested the fix so I'll answer.

`filter.toProto !== undefined` only checks if the object has a `toProto` properties (not set to `undefined`)
It is the case that `EntityFilter`s have a `toProto` methods but it is not necessarily the case that an object with a `toProto` property is an `EntityFilter`.

`filter instanceof EntityFilter` is the correct check for `filter is EntityFilter`

I hope that makes sense.

danieljbruce commented on src/query.ts:
```diff
@@ -207,12 +207,20 @@ class Query {
-    propertyOrFilter: string | EntityFilter,
-    operatorOrValue?: Operator,
-    value?: {} | null
+  filter(filter: EntityFilter): Query;
```
This signature change makes the input parameters more restrictive. Do we really want to go through with this? If so would this be a major version bump?

@vicb @bcoe

vicb replied:
My take is that the old signature was buggy and wouldn't catch unsupported calls. So I see this as a bug fix.

Also I think the impl could be improved by `switch`ing on `arguments.length`. 1, 2, 3 corresponding to the first 3 signatures in order.

danieljbruce replied:
I will do this in a separate issue:

https://github.com/googleapis/nodejs-datastore/issues/1096

ruyadorno reviewed (approved):
LGTM 👍 

Let's just make sure to keep an eye and monitor if the internal change from `src/filter.ts` `encodeValue()` is going to have any impact in the end users of the library. Releasing this as a patch version release is going to help us having the flexibility to rollback if needed, so I'd advise to proceed with a `fix:` commit.

Great work @danieljbruce!

danieljbruce committed:
Merge branch 'main' into 1065-query-filter-for-IN-with-keys-test

danieljbruce committed:
Merge branch 'main' into 1065-query-filter-for-IN-with-keys-test
